src/c4/C4_Functions.hh

```diff
@@ -352,8 +352,8 @@ template <typename T> void global_max(T &x);
 /*!
  * \brief Do an element-wise, global sum of an array.
  */
-template <typename T> void global_sum(T *x, int n);
-template <typename T> void global_sum(T *x, size_t n);
+template <typename T, typename L>
+void global_sum(T *x, L n, typename std::enable_if<std::is_integral<L>::value, L>::type * = 0);
```

Comment:
Oh man, this is hard to parse. I'm ok to merge this but we should probably have some comment around it to make it clear why the third argument is necessary.

Reply:
@along Good point.  Sorry - its a bit of a rush job.  Additionally, we should probably transform the remaining reductions in the same way.

Reply:
No problem! I like that it made the instantiations go away. I'm interested to learn more about `enable_if`.

Reply:
Do you prefer this form?

```c++
/*!
 * \brief Do an element-wise, global sum of an array.
 * 
 * Only instantiate this function for types L that integral.  Details of this pattern are at
 * https://en.cppreference.com/w/cpp/types/enable_if.
 */
template <typename T, typename L,
          typename std::enable_if<std::is_integral<L>::value, bool>::type = true>
void global_sum(T *x, L n);
````